fuel-vm/src/constraints/reg_key.rs

```diff
@@ -0,0 +1,312 @@
+    write_registers: &ProgramRegistersRef<'_>,
+) -> [Word; VM_REGISTER_COUNT] {
+    let mut out = [0u64; VM_REGISTER_COUNT];
+    out[..VM_REGISTER_READ_COUNT].copy_from_slice(&<[Word; VM_REGISTER_READ_COUNT]>::from(read_registers));
+    out[VM_REGISTER_READ_COUNT..].copy_from_slice(write_registers.0);
+    out
+}
+
+impl<'a> From<&'a SystemRegisters<'_>> for SystemRegistersRef<'a> {
+    fn from(value: &'a SystemRegisters<'_>) -> Self {
+        Self {
+            zero: Reg(value.zero.0),
+            one: Reg(value.one.0),
+            of: Reg(value.of.0),
+            pc: Reg(value.pc.0),
+            ssp: Reg(value.ssp.0),
+            sp: Reg(value.sp.0),
+            fp: Reg(value.fp.0),
+            hp: Reg(value.hp.0),
+            err: Reg(value.err.0),
+            ggas: Reg(value.ggas.0),
+            cgas: Reg(value.cgas.0),
+            bal: Reg(value.bal.0),
+            is: Reg(value.is.0),
+            ret: Reg(value.ret.0),
+            retl: Reg(value.retl.0),
+            flag: Reg(value.flag.0),
+        }
+    }
+}
+
+impl<'a> From<SystemRegisters<'a>> for SystemRegistersRef<'a> {
+    fn from(value: SystemRegisters<'a>) -> Self {
+        Self {
+            zero: Reg(value.zero.0),
+            one: Reg(value.one.0),
+            of: Reg(value.of.0),
+            pc: Reg(value.pc.0),
+            ssp: Reg(value.ssp.0),
+            sp: Reg(value.sp.0),
+            fp: Reg(value.fp.0),
+            hp: Reg(value.hp.0),
+            err: Reg(value.err.0),
+            ggas: Reg(value.ggas.0),
+            cgas: Reg(value.cgas.0),
+            bal: Reg(value.bal.0),
+            is: Reg(value.is.0),
+            ret: Reg(value.ret.0),
+            retl: Reg(value.retl.0),
+            flag: Reg(value.flag.0),
+        }
+    }
+}
+
+impl<'a> From<&'a ProgramRegisters<'_>> for ProgramRegistersRef<'a> {
+    fn from(value: &'a ProgramRegisters<'_>) -> Self {
+        Self(value.0)
+    }
+}
+
+impl<'a> From<ProgramRegisters<'a>> for ProgramRegistersRef<'a> {
+    fn from(value: ProgramRegisters<'a>) -> Self {
+        Self(value.0)
+    }
+}
+
+impl<'a> From<&SystemRegistersRef<'a>> for [Word; VM_REGISTER_READ_COUNT] {
+    fn from(value: &SystemRegistersRef<'a>) -> Self {
+        [
+            *value.zero,
+            *value.one,
+            *value.of,
+            *value.pc,
+            *value.ssp,
+            *value.sp,
+            *value.fp,
+            *value.hp,
+            *value.err,
+            *value.ggas,
+            *value.cgas,
+            *value.bal,
+            *value.is,
+            *value.ret,
+            *value.retl,
+            *value.flag,
+        ]
+    }
+}
+
+#[test]
+fn can_split() {
+    let mut reg: [Word; VM_REGISTER_COUNT] = std::iter::successors(Some(0), |x| Some(x + 1))
+        .take(VM_REGISTER_COUNT)
+        .collect::<Vec<_>>()
+        .try_into()
+        .unwrap();
+    let expect = reg;
+
+    let (r, w) = split_registers(&mut reg);
+
```

Comment:
Could you use `let SystemRegistersRef { and your variables here } = r;` please? It will help not to miss new registers=)